File: 90-SubsetsIi/90-SubsetsIi.java
// Last updated: 7/2/2025, 11:54:17 PM
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        help(res , nums, 0, new ArrayList<Integer>());
        return res;
    }
    public void help(List<List<Integer>> res, int[] nums, int i, ArrayList<Integer> lcl)
    {
        res.add(new ArrayList<Integer>(lcl));
        if(i == nums.length) return;

        for(int j=i;j<nums.length;j++)
        {
            if(i<j && nums[j] == nums[j-1]) continue;

            lcl.add(nums[j]);
            help(res , nums , j+1,lcl);
            lcl.remove(lcl.size()-1);
            //help(res, nums, j+1, lcl);
        }
    }

}

File: 90-SubsetsIi/90-SubsetsIi.cpp
// Last updated: 7/2/2025, 11:54:13 PM
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> lcl;

        sort(nums.begin(), nums.end());

        help(nums, ans, lcl, 0);

        return ans;
    }
      void help(vector<int>& nums, vector<vector<int>> &ans, vector<int>& lcl, int idx){
        if(idx == nums.size()+1) return;
        ans.push_back(lcl);

        for(int i = idx ; i < nums.size(); i++)
        {
            if(i > idx && nums[i] == nums[i-1]) continue;
            lcl.push_back(nums[i]);
            help(nums, ans, lcl ,i+1);
            lcl.pop_back();

        }
      }
};